plugin: new interface

Exirel opened the pull request:
I was playing for a long time with the idea of a unified Sopel's Plugin interface: a way to abstract how plugins are loaded from the core, so that you can implement various way of finding and loading them without having to change how the bot behave and interact with commands, rules, events, etc.

This PR **slightly** changes how callables are registered, and totally change how they are reloaded.

### Plugin Interface

The Plugin Interface is represented by the `sopel.plugins.handlers.AbstractPluginHandler`:

* `plugin.load()`: load the plugin in `sys.modules`
* `plugin.reload()`: reload the plugin
* `plugin.configure(settings)`: handle the configuration wizard for the plugin (if necessary)
* `plugin.setup(bot)`: setup the plugin (if necessary)
* `plugin.register(bot)`: register the plugin with the bot
* `plugin.shutdown(bot)`: take action on shutdown (if necessary)
* `plugin.is_loaded()`: tell if the plugin is loaded
* `plugin.has_*()` (configure, setup, shutdown): tell if the plugin has this particular method

Under the hood, I kept a lot of what is in `sopel.loader`, so it isn't a change that big!

### What's next?

I would like to add this kind of plugin:

* Python setup.py Entry Point based plugins: a Python package could define an entry point, and that would be used to discover plugins much more efficiently than the old technique of package namespace (as used by `sopel_modules.*`),
* Python object plugins: an handler that would not take a python module, but an object, to gather its callable, its configure, setup, and shutdown methods,

but that'll be for another day!

### Related issues, PR, and future works

* #1471 is about `sopel-*` CLI tools: this interface will be very useful to inspect plugins, show their details, configure and enable/disable them (as in #1519)
*  #1543 is about `time.sleep()`: with the rework of the JobScheduler and the new cycle of "shutdown, reload, register" that is better, it makes the JobScheduler and the `interval` decorator a good option for some cases (as requested by #1581 or seen in #831)

Exirel commented:
This PR is kind of related to #1471, as it modifies Sopel's internals, and it will definitively makes CLI tools easier to make.

For instance, it'll be much easier to implement a "list all plugins" CLI tool, or "configure this module in particular", etc.

Exirel commented:
Meh. Travis found a kind of cyclic import. Fine. I'll have to fix that one way or another. That'll be for tomorrow!

Exirel commented:
As said on IRC:

> 11:16:32 <Exirel> Yup, indeed, I forgot about `config.core.extra`. x)

So I need to update `enumerate_plugins` to load plugins from these directories ("extra" is a list).

Exirel commented:
As said on IRC: I merged #1490 in, so this PR will handle *everything* related to plugin's loading & reloading.

Also: so many things to fix, make, improve, and else. :smile:

Exirel commented:
I've figure out a way to:

* remove a circular import problem between sopel.loader, sopel.config, and sopel.plugins
* reduce the usage of sopel.loader to a very minimal set of functions
* put everything into sopel.cli.run and sopel.plugins instead

but I didn't add tests for sopel.plugins, which is **not good**, and I need to fix that.

And then, eventually, I'll ask for a review!

dgw commented:
Not dug into this yet because it's still tagged WIP, but does this basically replace #990, #1053, and possibly other PRs that deal with module loading/unloading/reloading?

Exirel commented:
Yup, it does. It's not perfect, because I do not handle (yet) sub-module reloading, but that'll come in later.

Exirel commented on docs/source/plugin.rst:
```diff
@@ -5,7 +5,12 @@ A Sopel plugin consists of a Python module containing one or more
+.. py:function:: callable(bot, trigger)
+
+    :param bot: the bot's instance
+    :type bot: :class:`sopel.bot.Sopel`
```
This is not true. As I discovered earlier, commands are called with an instance of `SopelWrapper`, and not `Sopel`. But shince `SopelWrapper` is not yet available from outside `Sopel`... it's hard to link it in the documentation.

This need to be fixed once this change is done.

Exirel replied:
Ok, so now it needs #1542 to be merged so I can change that.

Exirel commented:
So, for testing purpose, I need #1557 to be merged: I want to test the startup routine of the bot, when it loads plugins. And for that, I need to be able to stop the JobScheduler, otherwise the test suite never stop running. That could be annoying.

Exirel commented:
Yeah good I did push the "DO NOT PUSH" commits. Great. Have to stop Travis now. :(

Exirel commented:
Ok, I decorated these tests with `@pytest.mark.skip` in order to publish the code here. I had to rebase multiple times with the latest PR merged into master after 6.6.6 got released.

@dgw I think it's safe for you to start a review here. I'm sure you'll catch a lot of things to change or to discuss. It's been harder and harder to follow the change in master lately, and PR #1557 or alike will make it harder and harder.

2 things to consider:

1. Reloading is in a better shape than before, yet still has some issues, in particular with package plugins (as in `sopel_modules.plugin_name.sub_package`). For that, I'm thinking of a hook for plugin to reload themselves, like `def reload(bot)`.
2. I'm not sure package plugins actually works when put in `~/.sopel/modules` or into extra folders-that's something I need to check. Nonetheless, I think it's time to consider adding these folders to the Python Path automatically.

So, your review is very welcome at this point. If we can go through this one, we can implement many new features and improve on what already exists.

Exirel commented on sopel/plugins/__init__.py:
```diff
@@ -0,0 +1,188 @@
+import itertools
+import os
+
+from . import exceptions, handlers  # noqa
```
The lack of usage of `handlers` IIRC.

dgw replied:
Why import something that's not used?

Exirel replied:
So, it was the `exceptions` sub-module and not `handlers`.

The reason is to allow this:

```
from sopel import plugins

try:
    # something with plugins
except plugins.exceptions.PluginNotRegister:
    print('Oh no!')
```

Exirel commented on sopel/plugins/handlers.py:
```diff
@@ -0,0 +1,329 @@
+        :return: A human readable label for display purpose
+        :rtype: str
+
+        This method should, at least, return ``module_name + S + "module"``.
```
This is how the XML norm refers to whitespace, so your know that `module_name` is a variable, `S` is a space, and `"module"` is the string "module":

Valid: `admin module`
Invalid: `adminmodule`

Exirel commented on sopel/plugins/handlers.py:
```diff
@@ -0,0 +1,329 @@
+            return default_label
+
+        lines = inspect.cleandoc(module_doc).splitlines()
+        return default_label if not lines else lines[0]
```
I think I kept as much as I could the original logic here.

dgw replied:
Original from where? I'm having trouble finding where you might have adapted this from.

Exirel replied:
Found it, it's from there: https://github.com/sopel-irc/sopel/blob/v6.6.6/sopel/config/__init__.py#L199-L205

Exirel commented on sopel/plugins/handlers.py:
```diff
@@ -0,0 +1,329 @@
+        super(PyFilePlugin, self).__init__(name)
+
+    def _load(self):
+        # The current implementation uses the `imp.load_module` to perform
```
Don't be sorry. I'm glad you take time to fix my messy broken English.

Exirel commented on sopel/plugins/handlers.py:
```diff
@@ -0,0 +1,329 @@
+        # However, the solution for Python 3 is non-trivial, since the
+        # `importlib` built-in module does not have a similar function,
+        # therefore requires to dive into its public internals
+        # (``importlib.machinery`` and ``importlib.util``).
```
Yeah, it's complicated. I've to discuss that further, but it's clearly part of why I now want to add plugin's directories to the Python Path-something I was against at first, if you remember.

dgw replied:
"More to come", then. :D

btw, this `-` is a hyphen, not a dash (here you want `—`, em-dash). I noticed you using hyphens this way a few times recently, both in code and on IRC, so it was worth bringing up.

Exirel replied:
Welp, I've a long journey ahead of me to properly write in English!

Exirel commented on sopel/plugins/handlers.py:
```diff
@@ -0,0 +1,329 @@
+        ``reload`` function (either from `imp` or `importlib`), because the
+        module might not be available through ``sys.path``.
+        """
+        self._module = self._load()
```
Exactly.

Exirel commented on sopel/bot.py:
```diff
@@ -178,48 +179,120 @@ def write(self, args, text=None):  # Shim this in here for autodocs
+
+        total = sum([load_success, load_error, load_disabled])
+        if total and load_success:
+            stderr('\n\nRegistered %d modules,' % (load_success - 1))
```
Because it did before. Will remove it.

Exirel commented on sopel/plugins/__init__.py:
```diff
@@ -0,0 +1,188 @@
+def find_directory_plugins(directory):
+    """List plugins from a ``directory``
+
+    :return: Yield instance of :class:`~.handlers.PyFilePlugin`
```
I prefer the inline version for built-in types, so it'll be `:param str directory:`!

Exirel commented on sopel/plugins/__init__.py:
```diff
@@ -0,0 +1,188 @@
+    # load from directories
+    source_dirs = [os.path.join(settings.homedir, 'modules')]
+    if settings.core.extra:
+        source_dirs = source_dirs + list(settings.core.extra)
```
Good question. It's an habit of mine, I guess. I almost never use `append` or `extend`, because I don't like mutability.

Exirel commented on sopel/tools/jobs.py:
```diff
@@ -90,6 +90,19 @@ def clear_jobs(self):
             self._cleared = True
             self._jobs = PriorityQueue()
 
+    def remove_callable_job(self, callable):
```
This might produce merge conflict with #1557

Exirel replied:
It didn't for some reason... I just had to adapt the code afterward.

Exirel commented on sopel/cli/run.py:
```diff
@@ -244,6 +243,106 @@ def build_parser():
+
+def _plugins_wizard(settings):
+    usable_plugins = plugins.get_usable_plugins(settings)
+    for plugin, is_enabled in usable_plugins.items():
```
Nice catch. I think I replaced this too quickly to realize my mistake.

Exirel commented on sopel/plugins/__init__.py:
```diff
@@ -0,0 +1,188 @@
+
+def _list_plugin_filenames(directory):
+    # list plugin filenames from a directory
+    # yield 2-values tuples: (name, absolute path)
```
That's a mistake I make a lot. I can't remember when I should use plural or singular. :(

Exirel commented on sopel/plugins/handlers.py:
```diff
@@ -0,0 +1,329 @@
+    def load(self):
+        """Load the plugin
+
+        This method must be called first, in order to setup, register, shutdown
```
Best gang.

Exirel commented on sopel/cli/run.py:
```diff
@@ -244,6 +243,106 @@ def build_parser():
+        raise
+
+    name, ext = os.path.splitext(filename)
+    if not ext:
```
Fixed by doing some magic with `os.path.split`, which returns `(dirname, basename)`, where `basename` will be empty if the `filename` ends with a file path separator.

Exirel commented:
@dgw I fixed all the spelling mistakes, added a bunch of other fix here and there. I tested it a bit locally, just to be sure. I think this PR is in good shape, and is available for a new round of reviews.

In particular, I think it needs to be manually tested against a server, and play with a bit. However, at this point, I don't want to add new feature or new behavior.

I'm thinking about the `reload` feature in particular: it's less broken than before, because:

* URL callbacks are properly added/removed,
* Jobs are properly added/removed,
* shutdown & setup actions are now properly performed on reload,

but it still has issues with sub-module reloads. That's a feature I'd like to add/fix in another PR, by adding a hook instead of trying to be too smart with it. Another one is the question "should we add plugin directories to the Python Path?", that I think is worth another PR.

Exirel commented:
So, I've been running the bot for an hour now, trying a lot of commands, even trying reload! and so far, it works fine. I also try to play with the `sopel configure` command line, and it works fine too.

Pretty happy with that so far!

HumorBaby commented on sopel/plugins/exceptions.py:
```diff
@@ -0,0 +1,12 @@
+    pass
+
+
+class PluginNotRegistered(PluginError):
```
I think this should be:
```suggestion
        message = 'Plugin "%s" not registered' % name
```

HumorBaby commented on sopel/plugins/exceptions.py:
```diff
@@ -0,0 +1,12 @@
+
+class PluginNotRegistered(PluginError):
+    def __init__(self, name):
+        message = 'Plugin "%s" not registered'
```
I think this should be:
```suggestion
        message = 'Plugin "%s" not registered' % name
```

Exirel commented:
@HumorBaby oh my what a dumb mistake of mine! Fixed, thank you very much!

@dgw as said on IRC, I can squash most of the first commits: in the first 12, 11 can be squashed together, and the last one should stay alone (it's about documentation). I'm sure there are other commits that can be merged too.

dgw commented:
I'm not really bothered by those. I like making a commit for each step along the way toward building/rewriting something just as much as you do.

The fix/tweak commits in the middle are less appealing, but there's not really any good way to squash them in without throwing away a fair bit of metadata granularity re: who co-authored what. It's probably fine to leave the history as it is, since this PR was so long in the tooth getting ready.

Exirel commented on sopel/bot.py:
```diff
@@ -178,48 +179,120 @@ def write(self, args, text=None):  # Shim this in here for autodocs
+                    if func == self.memory['url_callbacks'].get(regex):
+                        self.unregister_url_callback(regex)
+        except:  # noqa
+            # TODO: consider logging?
```
I want to rework what we log in the whole setup/load/register/unregister/shutdown process. So that'll be for another PR.

HumorBaby commented on sopel/cli/run.py:
```diff
@@ -533,18 +520,22 @@ def command_legacy(opts):
         print_version()
         return
 
+    # TODO: allow to use a different homedir
```
Just a question: do you mean that you should be able to specify another directory to search for config files from the command line? 

If yes, nothing further :smile: 

If not, then I think this needs some clarification/discussion as to what is actually planned. I just wanted to confirm it's nothing will end up breaking without this `TODO`. The few runs I did to test various interpretations of this message seemed OK so far, so I'm not too worried :ok_hand:

Anyway, this is related to the CLI/wizard, and not the plugin system, so it doesn't preclude approval :+1:  I put it here since it wasn't present until now (that I can tell).

Exirel replied:
Yeah, we discussed on IRC that we should probably have a ``--homedir`` option to define the homedir at runtime, so you can do that:

    sopel start --homedir=/etc/sopel -c freenode

in order to load the configuration file `/etc/sopel/freenode.cfg` or to do that:

    sopel-config list --homedir=/etc/sopel

to list configuration files available in `/etc/sopel`. The problem is visible 2 lines below when we use `utils.find_config(config.DEFAULT_HOMEDIR, ...)`, and also with the config wizard.

HumorBaby replied:
Oh good. I did see the `utils.find_config(...` and was able to piece it together, but was initially  worried the `TODO` was related to the `core.homedir` setting. I quickly figured out that it was not actually the same, but it brings me to the question of whether `homedir` is the right term for that specific reference (and even `core.homedir` seems a little misnomer-y). Anyway, that can be dealt with in the appropriate context  when the time comes :grin:

Exirel commented:
> I ended up coming across some issues while testing this PR, but it's all related to the CLI

What kind of issues? 😮

dgw commented:
I trust @HumorBaby to open that other issue when he's ready. Unlike some of us (\*cough\* me \*cough\*), he's good at remembering what he promised to do. :grin:
